File: aoc15/ocaml/day_24/day_24.ml
let pkgs = Input.parse_lines int_of_string "input" in
let weight group = List.fold_left ( + ) 0 group in
let entanglement group = List.fold_left ( * ) 1 group in

let balance groups_cnt =
  let target = weight pkgs / groups_cnt in
  let rec power_set set remaining =
    let valid subset =
      List.length subset <= 8
      && weight subset <= target
      && weight subset + weight remaining >= target
    in
    match set with
    | [] -> [ [] ]
    | n :: ns ->
        let pns = power_set ns (n :: remaining) in
        List.filter valid (pns @ List.map (fun ss -> n :: ss) pns)
  in
  let groups = List.filter (fun g -> weight g = target) (power_set pkgs []) in
  let compare_groups g1 g2 =
    compare (List.length g1, entanglement g1) (List.length g2, entanglement g2)
  in
  let groups = List.sort compare_groups groups in
  entanglement (List.hd groups)
in

let first = balance 3 in
Printf.printf "First: %d\n" first;
let second = balance 4 in
Printf.printf "Second: %d\n" second

File: aoc15/go/day_24/day_24.go
package main

import (
	"aoc/collections"
	"bufio"
	"cmp"
	"fmt"
	"os"
	"slices"
	"strconv"

	"gonum.org/v1/gonum/stat/combin"
)

func weight(group []int, pkgs []int) (res int) {
	for _, g := range group {
		res += pkgs[g]
	}
	return
}

func entanglement(group []int, pkgs []int) int {
	res := 1
	for _, g := range group {
		res *= pkgs[g]
	}
	return res
}

func split(pkgs []int, n int) int {
	target := collections.Sum(pkgs) / n

	candidates := [][]int{}
	for k := 1; true; k++ {
		groups := combin.Combinations(len(pkgs), k)
		candidates = collections.Filter(groups, func(g []int) bool {
			return weight(g, pkgs) == target
		})
		if len(candidates) > 0 {
			break
		}
	}

	result := slices.MinFunc(candidates, func(a, b []int) int {
		return cmp.Compare(entanglement(a, pkgs), entanglement(b, pkgs))
	})

	return entanglement(result, pkgs)
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	pkgs := []int{}
	for scanner.Scan() {
		p, _ := strconv.Atoi(scanner.Text())
		pkgs = append(pkgs, p)
	}
	fmt.Println("First:", split(pkgs, 3))
	fmt.Println("Second:", split(pkgs, 4))
}
